Yaml parser errors out on deeply nested pipeline yaml file

### Component

server

### Describe the bug

When using a pipeline yaml file with deep nesting, Woodpecker shows a `max depth reached` error. Here is an example pipeline.yaml which fails reliably with this error:

```yaml
image: hello-world
when:
  - branch:
    - tester
steps:
  test:
    image: golang
    commands:
      - go install
      - go test
    backend_options:
      kubernetes:
        affinity:
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: accelerator
                      operator: In
                      values:
                        - nvidia-tesla-v100
```
The `backend_options` is taken directly from the [docs](https://woodpecker-ci.org/docs/administration/configuration/backends/kubernetes#example-node-affinity-for-gpu-workloads).

To me, it looks like the issue comes from the [xyaml lib](https://codeberg.org/6543/xyaml/src/branch/main/merge.go).

### Steps to reproduce

I've added a UT for the `ParseString` function: https://github.com/woodpecker-ci/woodpecker/blob/db70e65d21f8daa84808c6303d9608e4f72908bd/pipeline/frontend/yaml/parse.go#L34

See this commit in my fork: https://github.com/woodpecker-ci/woodpecker/commit/ef741be81eb51664b5ae0e26aaac4d9dc179b877

It runs the example Yaml I've shown above. The UT is executed when running `make test-lib` and fails with this error:
```
--- FAIL: TestParse (0.01s)
    --- FAIL: TestParse/Should_handle_deeply_nested_yaml (0.00s)
        parse_test.go:79: 
                Error Trace:    /home/michael/projects/woodpecker/pipeline/frontend/yaml/parse_test.go:79
                Error:          Received unexpected error:
                                max depth reached
                Test:           TestParse/Should_handle_deeply_nested_yaml
FAIL
coverage: 100.0% of statements
FAIL    go.woodpecker-ci.org/woodpecker/v3/pipeline/frontend/yaml       0.035s
```

### Expected behavior

I expect parsing of the example Yaml file to not fail.

### System Info

```shell
My version is 3.13.0:

{"source":"https://github.com/woodpecker-ci/woodpecker","version":"3.13.0"}
```

### Additional context

_No response_

### Validations

- [x] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [x] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [x] Checked that the bug isn't fixed in the `next` version already [https://woodpecker-ci.org/versions]

--- a/pipeline/frontend/yaml/parse.go
+++ b/pipeline/frontend/yaml/parse.go
@@ -21,7 +21,7 @@ import (
 )
 
 const (
-	// Set to allow deeply nested k8s backend_options for node affinity
+	// Set to allow deeply nested k8s backend_options for node affinity.
 	maxMergeDepth uint16 = 15
 )
 

--- a/pipeline/frontend/yaml/parse.go
+++ b/pipeline/frontend/yaml/parse.go
@@ -20,10 +20,16 @@ import (
 	"go.woodpecker-ci.org/woodpecker/v3/pipeline/frontend/yaml/types"
 )
 
+const (
+	// Set to allow deeply nested k8s backend_options for node affinity
+	maxMergeDepth uint16 = 15
+)
+
 // ParseBytes parses the configuration from bytes b.
 func ParseBytes(b []byte) (*types.Workflow, error) {
+	parser := xyaml.NewParser(xyaml.WithDepth(maxMergeDepth))
 	out := new(types.Workflow)
-	err := xyaml.Unmarshal(b, out)
+	err := parser.Unmarshal(b, out)
 	if err != nil {
 		return nil, err
 	}

--- a/pipeline/frontend/yaml/parse_test.go
+++ b/pipeline/frontend/yaml/parse_test.go
@@ -77,6 +77,11 @@ func TestParse(t *testing.T) {
 		assert.Equal(t, "plugins/slack", out.Steps.ContainerList[1].Image)
 		assert.Equal(t, yaml_base_types.StringOrSlice{"push"}, out.Steps.ContainerList[1].When.Constraints[0].Event)
 	})
+
+	t.Run("Should handle deeply nested yaml", func(t *testing.T) {
+		_, err := ParseString(sampleDeepYaml)
+		assert.NoError(t, err)
+	})
 }
 
 func TestMatch(t *testing.T) {
@@ -243,6 +248,30 @@ steps:
     environment:
       DRIVER: next
       PLATFORM: linux
+`
+
+var sampleDeepYaml = `
+image: hello-world
+when:
+  - branch:
+    - tester
+steps:
+  test:
+    image: golang
+    commands:
+      - go install
+      - go test
+    backend_options:
+      kubernetes:
+        affinity:
+          nodeAffinity:
+            requiredDuringSchedulingIgnoredDuringExecution:
+              nodeSelectorTerms:
+                - matchExpressions:
+                    - key: accelerator
+                      operator: In
+                      values:
+                        - nvidia-tesla-v100
 `
 
 func TestReSerialize(t *testing.T) {

